File: src/background/core-stream-parser.ts
export interface StreamedComponent {
  sentenceId: string;
  component: Record<string, unknown>;
}

interface Frame {
  kind: "object" | "array";
  /** The key this container sits under in its parent, when there is one. */
  key?: string;
  /** Object frames only: the key whose value is being read next. */
  pendingKey?: string;
}

interface SentenceFrame {
  /** Offset in the buffer where this sentence object opened. */
  start: number;
  sentenceId?: string;
  /** Components that arrived before this sentence's id did. */
  buffered: Array<Record<string, unknown>>;
}

const SENTENCE_ID_PATTERN = /"sentenceId"\s*:\s*"((?:[^"\\]|\\.)*)"/u;

/**
 * Pulls whole components out of a core-analysis envelope while it is still
 * streaming, so a paragraph can start rendering before the model has finished
 * the sentence.
 *
 * This is deliberately not a general streaming JSON parser: it tracks only the
 * one shape the core prompt asks for —
 * `{"sentences":[{"sentenceId":…,"components":[{…},…]}]}` — and emits each
 * component object once it closes. A component is recognised only inside an
 * array keyed `components` inside an element of the array keyed `sentences`;
 * objects nested within a component close first and are not mistaken for it.
 *
 * Callers must treat every emitted component as unvalidated model output. The
 * coverage invariant (every non-punctuation token covered exactly once) can only
 * be checked once the whole sentence has arrived, so these are display-only.
 */
export class CoreStreamParser {
  #stack: Frame[] = [];
  #sentences: SentenceFrame[] = [];
  #buffer = "";
  #cursor = 0;
  #inString = false;
  #escaped = false;
  #started = false;
  #stringStart = 0;
  /** Offset where the component object currently open began, if any. */
  #componentStart: number | undefined;

  /** Feeds more of the raw JSON text and returns the components it completed. */
  push(text: string): StreamedComponent[] {
    this.#buffer += text;
    const emitted: StreamedComponent[] = [];
    while (this.#cursor < this.#buffer.length) {
      const index = this.#cursor;
      const character = this.#buffer[index]!;
      this.#cursor += 1;

      if (this.#inString) {
        if (this.#escaped) this.#escaped = false;
        else if (character === "\\") this.#escaped = true;
        else if (character === '"') this.#inString = false;
        continue;
      }

      if (!this.#started) {
        // Tolerate a Markdown fence or any prose the model emits first.
        if (character !== "{") continue;
        this.#started = true;
      }

      switch (character) {
        case '"':
          this.#inString = true;
          this.#stringStart = index;
          break;
        case ":":
          this.#assignPendingKey();
          break;
        case "{":
          this.#openObject(index);
          break;
        case "[":
          this.#stack.push({ kind: "array", key: this.#consumeKey() });
          break;
        case "}":
          this.#closeObject(index, emitted);
          break;
        case "]":
          this.#stack.pop();
          break;
        default:
          break;
      }
    }
    return emitted;
  }

  /** A colon always follows a key literal, so the last closed string is it. */
  #assignPendingKey(): void {
    const frame = this.#stack.at(-1);
    if (frame?.kind !== "object") return;
    const literal = this.#buffer.slice(this.#stringStart, this.#cursor - 1).trim();
    frame.pendingKey = parseString(literal);
  }

  #consumeKey(): string | undefined {
    const frame = this.#stack.at(-1);
    if (frame?.kind !== "object") return undefined;
    const key = frame.pendingKey;
    frame.pendingKey = undefined;
    return key;
  }

  #openObject(index: number): void {
    const key = this.#consumeKey();
    const parent = this.#stack.at(-1);
    this.#stack.push({ kind: "object", key });
    if (parent?.kind !== "array") return;
    if (parent.key === "sentences") {
      this.#sentences.push({ start: index, buffered: [] });
      return;
    }
    if (parent.key === "components" && this.#componentStart === undefined) {
      this.#componentStart = index;
    }
  }

  #closeObject(index: number, emitted: StreamedComponent[]): void {
    if (this.#stack.pop() === undefined) return;
    const enclosing = this.#stack.at(-1);
    if (this.#componentStart !== undefined && enclosing?.key === "components") {
      const component = parseObject(this.#buffer.slice(this.#componentStart, index + 1));
      this.#componentStart = undefined;
      if (component !== undefined) this.#collect(component, emitted);
      return;
    }
    if (enclosing?.key === "sentences") this.#flushBuffered(emitted);
  }

  #collect(component: Record<string, unknown>, emitted: StreamedComponent[]): void {
    const sentence = this.#sentences.at(-1);
    if (sentence === undefined) return;
    sentence.sentenceId ??= this.#readSentenceId(sentence);
    if (sentence.sentenceId === undefined) {
      // The model put components before the id; hold them until it arrives.
      sentence.buffered.push(component);
      return;
    }
    emitted.push({ sentenceId: sentence.sentenceId, component });
  }

  /** Scans only this sentence's own text, so sibling sentences cannot bleed in. */
  #readSentenceId(sentence: SentenceFrame): string | undefined {
    const match = this.#buffer.slice(sentence.start).match(SENTENCE_ID_PATTERN);
    return match === null ? undefined : parseString(`"${match[1]!}"`);
  }

  #flushBuffered(emitted: StreamedComponent[]): void {
    const sentence = this.#sentences.at(-1);
    if (sentence === undefined || sentence.buffered.length === 0) return;
    const sentenceId = sentence.sentenceId ?? this.#readSentenceId(sentence);
    if (sentenceId === undefined) return;
    for (const component of sentence.buffered) emitted.push({ sentenceId, component });
    sentence.buffered = [];
  }
}

function parseString(literal: string): string | undefined {
  if (!literal.startsWith('"') || !literal.endsWith('"') || literal.length < 2) return undefined;
  try {
    const value: unknown = JSON.parse(literal);
    return typeof value === "string" ? value : undefined;
  } catch {
    return undefined;
  }
}

function parseObject(text: string): Record<string, unknown> | undefined {
  try {
    const value: unknown = JSON.parse(text);
    return typeof value === "object" && value !== null && !Array.isArray(value)
      ? (value as Record<string, unknown>)
      : undefined;
  } catch {
    return undefined;
  }
}

File: intellij-plugin/src/main/kotlin/dev/codetui/englishsyntax/model/CoreStreamParser.kt
package dev.codetui.englishsyntax.model

import kotlinx.serialization.json.JsonObject

data class StreamedComponent(
  val sentenceId: String,
  val component: JsonObject,
)

private data class Frame(
  val kind: Kind,
  val key: String? = null,
  var pendingKey: String? = null,
) {
  enum class Kind { OBJECT, ARRAY }
}

private data class SentenceFrame(
  val start: Int,
  var sentenceId: String? = null,
  var buffered: MutableList<JsonObject> = mutableListOf(),
)

/**
 * 在核心分析 envelope 仍在流式输出时逐出完整 component，让段落提前渲染。
 * 刻意不是通用 JSON 解析器：只跟踪 prompt 要求的固定形状，并在 component 对象
 * 闭合时输出。component 是未校验的模型输出，调用方必须再做最终校验。
 */
class CoreStreamParser {
  private val stack = mutableListOf<Frame>()
  private val sentences = mutableListOf<SentenceFrame>()
  private val buffer = StringBuilder()
  private var cursor = 0
  private var inString = false
  private var escaped = false
  private var started = false
  private var stringStart = 0
  private var componentStart: Int? = null

  /** Feeds more raw JSON text and returns the components it completed. */
  fun push(text: String): List<StreamedComponent> {
    buffer.append(text)
    val emitted = mutableListOf<StreamedComponent>()
    while (cursor < buffer.length) {
      val index = cursor
      val character = buffer[index]
      cursor += 1

      if (inString) {
        when {
          escaped -> escaped = false
          character == '\\' -> escaped = true
          character == '"' -> inString = false
        }
        continue
      }

      if (!started) {
        // 容忍模型先吐 Markdown 围栏或散文。
        if (character != '{') continue
        started = true
      }

      when (character) {
        '"' -> {
          inString = true
          stringStart = index
        }
        ':' -> assignPendingKey()
        '{' -> openObject(index)
        '[' -> stack += Frame(Frame.Kind.ARRAY, key = consumeKey())
        '}' -> closeObject(index, emitted)
        ']' -> stack.removeLastOrNull()
      }
    }
    return emitted
  }

  private fun assignPendingKey() {
    val frame = stack.lastOrNull() ?: return
    if (frame.kind != Frame.Kind.OBJECT) return
    val literal = buffer.substring(stringStart, cursor - 1).trim()
    frame.pendingKey = parseStreamString(literal)
  }

  private fun consumeKey(): String? {
    val frame = stack.lastOrNull() ?: return null
    if (frame.kind != Frame.Kind.OBJECT) return null
    val key = frame.pendingKey
    frame.pendingKey = null
    return key
  }

  private fun openObject(index: Int) {
    val key = consumeKey()
    val parent = stack.lastOrNull()
    stack += Frame(Frame.Kind.OBJECT, key = key)
    if (parent?.kind != Frame.Kind.ARRAY) return
    when (parent.key) {
      "sentences" -> sentences += SentenceFrame(start = index)
      "components" -> if (componentStart == null) componentStart = index
    }
  }

  private fun closeObject(index: Int, emitted: MutableList<StreamedComponent>) {
    stack.removeLastOrNull() ?: return
    val enclosing = stack.lastOrNull()
    val start = componentStart
    if (start != null && enclosing?.key == "components") {
      val component = parseStreamObject(buffer.substring(start, index + 1))
      componentStart = null
      if (component != null) collect(component, emitted)
      return
    }
    if (enclosing?.key == "sentences") flushBuffered(emitted)
  }

  private fun collect(component: JsonObject, emitted: MutableList<StreamedComponent>) {
    val sentence = sentences.lastOrNull() ?: return
    if (sentence.sentenceId == null) {
      sentence.sentenceId = readSentenceId(sentence)
    }
    val sentenceId = sentence.sentenceId
    if (sentenceId == null) {
      sentence.buffered += component
      return
    }
    emitted += StreamedComponent(sentenceId, component)
  }

  private fun readSentenceId(sentence: SentenceFrame): String? {
    val match = SENTENCE_ID_PATTERN.find(buffer.substring(sentence.start)) ?: return null
    return parseStreamString("\"${match.groupValues[1]}\"")
  }

  private fun flushBuffered(emitted: MutableList<StreamedComponent>) {
    val sentence = sentences.lastOrNull() ?: return
    if (sentence.buffered.isEmpty()) return
    val sentenceId = sentence.sentenceId ?: readSentenceId(sentence) ?: return
    for (component in sentence.buffered) emitted += StreamedComponent(sentenceId, component)
    sentence.buffered = mutableListOf()
  }

  companion object {
    private val SENTENCE_ID_PATTERN = Regex("\"sentenceId\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"")
  }
}

